### PR TITLE
refactor!: license arguments spdx compliant

### PR DESCRIPTION
**Pull request by neodmy**
### Main changes

- :new: (feat)
- :wrench: (chore)
- :memo: (docs)
- :arrows_counterclockwise: (refactor)
- :microscope: (test)
- :bangbang: (breaking change)

This PR is a **BREAKING CHANGE**. The main goal of this PR is to keep the current functionality but making the scanning licenses process SPDX compliant. For that, I've done the following:

- Remove the utility functions related to the old approach for the matching process. We were using string or RegEx's passed to `failOn` to check if any of the packages' licenses matched the value. As we no longer need that, I've removed all the related functions and their tests
- Remove the `spdx-validator` component introduced in the previous PR. The only reason for this component to exist was the wrapper around the `spdx-expression-validate` to throw an error when the license wasn't SPDX compliant. This was used to check the value of `checkLicense` option. I've moved this code from the `index.js` to `runner.js`, and control the error there. So there is no need to keep the `spdx-validator` component anymore, we import the SPDX packages on the `runner.js` and use them
- `failOn` stays as a list and `generateOutputOn` has been changed to a list for consistency. I've updated its description and the type on the `args.js` file (the `yargs` setup), and the documentation.
- Remove the args' middleware. The old functions were used to verify if the RegEx values passed to `failOn` were valid, and to parse the `generateOutput` licenses. These were meant to be passed as a comma separated list of licenses, which no longer applies.
- Refactor the `runner` logic to use `spdx-validator` and `spdx-expression-validate` packages.
  - As mentioned before, the `runner` now controls the `checkLicenses` flow. If a value is passed, it will check it and finish without triggering the scanning process
  - The list of licenses passed to `failOn` and `generateOutputOn` options, if provided, must be a SPDX compliant. Otherwise, an error will be thrown. If all are correct, they are concatenated using the `OR` operand. The result of this operation is used for the matching process. So if executed with `--failOn MIT GPL-1.0 0BSD`, the SPDX expression to check the package licenses will be `MIT OR GPL-1.0 OR 0BSD`
  - Packages' licenses must also be SPDX. So if any of the `package.json` entries has a non-compliant license, it won't be taken into account for the scanning process. A warning message will be printed to inform what packages are in this situation. We could potentially include an option to fail if that's the case.

To sum up, the main functionality stays the same. However, the checking process will make use of the SPDX packages. This will provide support for the `AND`, `OR` and `WITH` syntax during the matching process, which previously wasn't possible. 

### Additional notes

README has been updated accordingly. We've found an issue in spdx-satisfies package. An explanation has been added to this document as well.
I have had some errors with eslint (incompatible packages). So I've reinstalled all the `eslint` dependencies and replaced the `browser` option by `node` on the `.eslintrc.json`

### Context

- :ticket: Closes / Relates #53

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -22,19 +22,19 @@ You can optionally add options to exclude generating the report or avoid generat
...
-  ```
+```sh
+npm install @guidesmiths/license-checker
+```
 
 - Add a script to run the package
 
-```
-npx @guidesmiths/license-checker --failOn license1 license2
+```sg
```
Comment (inigomarquinez) [inline README.md:31]: typo?

```suggestion
```sh
```

File: test/runner.test.js [hunk 2]
```diff
@@ -0,0 +1,325 @@
+const runner = require('../src/runner');
```
Comment (inigomarquinez) [inline test/runner.test.js:1]: AWESOME 👏🏼

File: src/runner.js [hunk 3]
```diff
@@ -1,26 +1,62 @@
+const isSPDXCompliant = require('spdx-expression-validate');
+const satisfiesSPDXLicense = require('spdx-satisfies');
+
 const {
-  getPackageInfoList, extractInvalidPackages, formatForbiddenLicenseError
+  getPackageInfoList, formatForbiddenLicenseError
 } = require('./utils');
 
-const reporter = require('./reporter');
+const checkLicenseArgument = (args = []) => {
```
Comment (inigomarquinez) [inline src/runner.js:8]: @neodmy , do you find it useful to move `checkLicenseArgument`, `generateSPDXExpression` and `checkPackagesLicenses` to the `utils` folder? Or is it better to keep them here?

**Review by inigomarquinez — COMMENTED** [review]
@neodmy , I think you've done an awesome work!! Thank you so much for your contribution.

I've added a couple of minor comments before approving

File: src/runner.js (hunk 3 above)
Comment (neodmy) [reply src/runner.js]: Done. I've renamed some functions and their arguments to make them more _agnostic_

**Review by inigomarquinez — APPROVED** [review]
LGTM!

**Comment by neodmy** [comment]
@nanotower is taking over this PR. As discussed, he will implement an alternative approach which will remove the licenses passed to `failOn` from the complete SPDX list. Then, will check whether every package satisfies at least one license of the resulting list.

File: src/utils.js [hunk 4]
```diff
@@ -82,9 +50,80 @@ const formatForbiddenLicenseError = licenses => {
...
+ */
+const isLicenseError = (license = '') => licensesExceptions.includes(license);
+
+/**
+ * Checks the SPDX expression against the list of packages. If the license of the package itself is not SPDX compliant, the package will
+ * be included on the "nonCompliant" list. If the SPDX expression satisfies the package license, the package will be included on the
+ * "satisfied" list
+ * @param {string} expression - A SPDX expression
+ * @param {object[]} packages - A list of packages to be checked against the SPDX expression
+ * @return {{satisfied: object[], nonCompliant: object[]}} - A couple of lists including the packages that satisfy the SPDX expression
```
Comment (neodmy) [inline src/utils.js:102]: We should double-check the JSDoc for this function after the latest changes. The return is now different, and probably the semantic of the function is too.

Comment (nanotower) [reply src/utils.js]: Done

**Review by inigomarquinez — APPROVED** [review]

**Comment by inigomarquinez** [comment]
@all-contributors please add @nanotower for code

**Comment by inigomarquinez** [comment]
@all-contributors please add @nanotower for code
